pkg/util/watch/delete.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watch
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// These constants are used for log field keys.
+// They mirror the keys used by the controller-manager.
+const (
+	logFieldObjectRef  = "object"
+	logFieldObjectKind = "objectKind"
+)
+
+// DeleteAndWait deletes the object (if not already deleted) and blocks until it
+// is fully deleted (NotFound, not just deleteTimestamp). This allows for any
+// finalizers to complete.
+//
+// Exits early if the object is already deleted or doesn't exist.
+//
+// The Object contents will be replaced with updated state for each new event
+// received while waiting for deletion.
+//
+// If you want to time out the full function, use a context.WithTimeout.
+// If you want to time out just the waiting after the delete, use reconcileTimeout.
+func DeleteAndWait(ctx context.Context, c client.WithWatch, obj client.Object, reconcileTimeout time.Duration) error {
+	gvk, err := kinds.Lookup(obj, c.Scheme())
+	if err != nil {
+		return err
+	}
+	id := core.ID{
+		ObjectKey: client.ObjectKeyFromObject(obj),
+		GroupKind: gvk.GroupKind(),
+	}
+
+	// Use cancellable context to delay starting the timer until after the
+	// watcher has synchronized and the delete call has completed.
+	watchCtx, cancel := context.WithCancel(ctx)
+	// Stop early, if done or errored before timeout.
+	defer cancel()
+	// Delay timer start until after Sync and Delete.
+	startTimer := func() {
+		timerCtx, cancel2 := context.WithTimeout(watchCtx, reconcileTimeout)
```

Comment:
Maybe something like `timeoutCancel` instead of `cancel2`?

Reply:
done